[BUG] Non-blocking make export error: missing gnu-elf.ld

### Description / Steps to reproduce the issue

When I configure for `stm32f4discovery:nsh` and then run `make export`, I get a warning about a missing linker script:
```
cp: cannot stat '/home/thomas/nuttxspace/nuttx/libs/libc/elf/gnu-elf.ld': No such file or directory
```

Despite this, the export archive is created successfully and the build is not blocked. It seems the export step is trying to copy a file that doesn’t exist anymore.

### Steps to Reproduce:

1. Configure NuttX:
```
./tools/configure.sh -l stm32f4discovery:nsh
```
2. Run export:
```
make export
```

**Output**:
```
Create version.h
LN: platform/board to /home/thomas/nuttxspace/apps/platform/dummy
Register: hello
Register: dd
Register: nsh
Register: sh
CC:  chip/stm32_gpio.c chip/stm32_gpio.c:44:11: note: '#pragma message: CONFIG_STM32_USE_LEGACY_PINMAP will be deprecated migrate board.h see tools/stm32_pinmap_tool.py'
   44 | #  pragma message "CONFIG_STM32_USE_LEGACY_PINMAP will be deprecated migrate board.h see tools/stm32_pinmap_tool.py"
      |           ^~~~~~~
CC:  chip/stm32_flash.c In file included from chip/stm32_flash.c:40:
chip/stm32f20xxf40xx_flash.c:53:4: warning: #warning "Default Flash Configuration Used - See Override Flash Size Designator" [-Wcpp]
   53 | #  warning "Default Flash Configuration Used - See Override Flash Size Designator"
      |    ^~~~~~~
cp: cannot stat '/home/thomas/nuttxspace/nuttx/libs/libc/elf/gnu-elf.ld': No such file or directory
```

*Notes*:
The error is non-blocking and does not prevent the export archive from being created. I just wanted to check if this is an intentional leftover or something that should be cleaned up. I’d be happy to look into it further if needed.

### On which OS does this issue occur?

[OS: Linux]

### What is the version of your OS?

Ubuntu 24.04

### NuttX Version

master

### Issue Architecture

[Arch: arm]

### Issue Area

[Area: Configuring]

### Host information

_No response_

### Verification

- [x] I have verified before submitting the report.

--- a/tools/mkexport.sh
+++ b/tools/mkexport.sh
@@ -19,6 +19,8 @@
 # under the License.
 #
 
+set -e   # Exit on error
+
 # Get the input parameter list
 
 USAGE="USAGE: $0 [-d] [-z] [-u] [-t <top-dir> [-x <lib-ext>] [-a <apps-dir>] [-m <make-exe>] -l \"lib1 [lib2 [lib3 ...]]\""
@@ -185,9 +187,10 @@ cp "${TOPDIR}/tools/incdir.c" "${EXPORTDIR}/tools/."
 # Copy the board specific linker if found, or use the default when not.
 
 APPLD=gnu-elf.ld
+
 if [ -f "${BOARDDIR}/scripts/${APPLD}" ]; then
   cp -f "${BOARDDIR}/scripts/${APPLD}" "${EXPORTDIR}/scripts/."
-else
+elif [ -f "${TOPDIR}/libs/libc/elf/${APPLD}" ]; then
   cp -f "${TOPDIR}/libs/libc/elf/${APPLD}" "${EXPORTDIR}/scripts/."
 fi
 
@@ -364,70 +367,101 @@ cp -LR -p "${TOPDIR}/include" "${EXPORTDIR}/." || \
 ${MAKE} -C ${ARCHDIR} export_startup TOPDIR=${TOPDIR} EXPORT_DIR="${EXPORTDIR}"
 
 # Copy architecture-specific header files into the arch export sub-directory.
-# This is tricky because each architecture does things in a little different
-# way.
-#
-# First copy any header files in the architecture src/ sub-directory (some
-# architectures keep all of the header files there, some a few, and others
-# none
+# Some architectures keep all headers in src/, some only a few, and others none.
 
-cp -f "${ARCHDIR}"/*.h "${EXPORTDIR}"/arch/. 2>/dev/null
-
-# Then look a list of possible places where other architecture-specific
-# header files might be found.  If those places exist (as directories or
-# as symbolic links to directories, then copy the header files from
-# those directories into the EXPORTDIR
+if [ -d "${ARCHDIR}" ]; then
+    # Expand the glob safely
+    set -- "${ARCHDIR}"/*.h
+    if [ -e "$1" ]; then
+        echo "MK: Copying architecture headers from ${ARCHDIR} to ${EXPORTDIR}/arch"
+        if ! cp -f "$@" "${EXPORTDIR}/arch/"; then
+            echo "MK: Error: Failed to copy headers from ${ARCHDIR}" >&2
+            exit 1
+        fi
+    else
+        echo "MK: Notice: No header files found in ${ARCHDIR}" >&2
+    fi
+else
+    echo "MK: Warning: Architecture directory ${ARCHDIR} does not exist" >&2
+fi
 
 if [ "X${USRONLY}" != "Xy" ]; then
   ARCH_HDRDIRS="arm armv7-m avr avr32 board common chip mips32"
+
   for hdir in $ARCH_HDRDIRS; do
+    srcdir="${ARCHDIR}/${hdir}"
+    dstdir="${EXPORTDIR}/arch/${hdir}"
 
     # Does the directory (or symbolic link) exist?
+    if [ -d "$srcdir" ] || [ -h "$srcdir" ]; then
+      mkdir -p "$dstdir" || {
+        echo "MK: Error: mkdir $dstdir failed" >&2
+        exit 1
+      }
 
-    if [ -d "${ARCHDIR}/${hdir}" -o -h "${ARCHDIR}/${hdir}" ]; then
+      # Copy headers if they exist
+      set -- "$srcdir"/*.h
+      if [ -e "$1" ]; then
+        echo "MK: Copying headers from $srcdir to $dstdir"
+        if ! cp -f "$@" "$dstdir/"; then
+          echo "MK: Error: Failed to copy headers from $srcdir" >&2
+          exit 1
+        fi
+      else
+        echo "MK: Notice: No header files found in $srcdir" >&2
+      fi
 
-      # Yes.. create a export sub-directory of the same name
+      # Handle hardware subdir if it exists
+      if [ -d "$srcdir/hardware" ]; then
+        mkdir -p "$dstdir/hardware" || {
+          echo "MK: Error: mkdir $dstdir/hardware failed" >&2
+          exit 1
+        }
 
-      mkdir "${EXPORTDIR}/arch/${hdir}" || \
-        { echo "MK: 'mkdir ${EXPORTDIR}/arch/${hdir}' failed"; exit 1; }
-
-      # Then copy the header files (only) into the new directory
-
-      cp -f "${ARCHDIR}"/${hdir}/*.h "${EXPORTDIR}"/arch/${hdir}/. 2>/dev/null
-
-      # Most architectures have low directory called "hardware" that
-      # holds the header files
-
-      if [ -d "${ARCHDIR}/${hdir}/hardware" ]; then
-
-        # Yes.. create a export sub-directory of the same name
-
-        mkdir "${EXPORTDIR}/arch/${hdir}/hardware" || \
-          { echo "MK: 'mkdir ${EXPORTDIR}/arch/${hdir}/hardware' failed"; exit 1; }
-
-        # Then copy the header files (only) into the new directory
-
-        cp -f "${ARCHDIR}"/${hdir}/hardware/*.h "${EXPORTDIR}"/arch/${hdir}/hardware/. 2>/dev/null
+        set -- "$srcdir/hardware"/*.h
+        if [ -e "$1" ]; then
+          echo "MK: Copying headers from $srcdir/hardware to $dstdir/hardware"
+          if ! cp -f "$@" "$dstdir/hardware/"; then
+            echo "MK: Error: Failed to copy headers from $srcdir/hardware" >&2
+            exit 1
+          fi
+        else
+          echo "MK: Notice: No hardware headers in $srcdir/hardware" >&2
+        fi
       fi
     fi
   done
 
-  # Copy OS internal header files as well.  They are used by some architecture-
-  # specific header files.
-
-  mkdir "${EXPORTDIR}/arch/os" || \
-    { echo "MK: 'mkdir ${EXPORTDIR}/arch/os' failed"; exit 1; }
+  # Copy OS internal header files
+  mkdir -p "${EXPORTDIR}/arch/os" || {
+    echo "MK: Error: mkdir ${EXPORTDIR}/arch/os failed" >&2
+    exit 1
+  }
 
   OSDIRS="clock environ errno group init irq mqueue paging pthread sched semaphore signal task timer wdog"
 
   for dir in ${OSDIRS}; do
-    mkdir "${EXPORTDIR}/arch/os/${dir}" || \
-      { echo "MK: 'mkdir ${EXPORTDIR}/arch/os/${dir}' failed"; exit 1; }
-    cp -f "${TOPDIR}"/sched/${dir}/*.h "${EXPORTDIR}"/arch/os/${dir}/. 2>/dev/null
+    srcdir="${TOPDIR}/sched/${dir}"
+    dstdir="${EXPORTDIR}/arch/os/${dir}"
+
+    mkdir -p "$dstdir" || {
+      echo "MK: Error: mkdir $dstdir failed" >&2
+      exit 1
+    }
+
+    set -- "$srcdir"/*.h
+    if [ -e "$1" ]; then
+      echo "MK: Copying OS headers from $srcdir to $dstdir"
+      if ! cp -f "$@" "$dstdir/"; then
+        echo "MK: Error: Failed to copy headers from $srcdir" >&2
+        exit 1
+      fi
+    else
+      echo "MK: Notice: No headers in $srcdir" >&2
+    fi
   done
 
-  # Add the board library to the list of libraries
-
+  # Add the board library if present
   if [ -f "${ARCHDIR}/board/libboard${LIBEXT}" ]; then
     LIBLIST="${LIBLIST} ${ARCHSUBDIR}/board/libboard${LIBEXT}"
   fi

--- a/tools/toolchain.cmake.export
+++ b/tools/toolchain.cmake.export
@@ -19,20 +19,17 @@ set(CMAKE_CXX_STANDARD_INCLUDE_DIRECTORIES
     ${NUTTX_PATH}/include/${NUTTX_CXX} ${NUTTX_PATH}/include
     ${NUTTX_PATH}/arch/chip)
 
-file(GLOB STARTUP_OBJS ${NUTTX_PATH}/startup/*)
-
 add_compile_options(-nostdlib)
 add_compile_options(-ffunction-sections -fdata-sections)
 
 # same entry used for all build modes in crt0.c and arch/.../xxx_start.c
-
 set(ENTRY_NAME "__start")
 
 set(CMAKE_C_LINK_EXECUTABLE
-    "<CMAKE_LINKER> ${LDFLAGS} --entry=${ENTRY_NAME} -T${LINKER_SCRIPT} <OBJECTS> ${STARTUP_OBJS} -o <TARGET> <LINK_LIBRARIES> -L${NUTTX_PATH}/libs --start-group ${LDLIBS} ${EXTRA_LIBS} --end-group"
+    "<CMAKE_LINKER> ${LDFLAGS} --entry=${ENTRY_NAME} -T${LINKER_SCRIPT} <OBJECTS> -o <TARGET> <LINK_LIBRARIES> -L${NUTTX_PATH}/libs --start-group ${LDLIBS} ${EXTRA_LIBS} --end-group"
 )
 set(CMAKE_CXX_LINK_EXECUTABLE
-    "<CMAKE_LINKER> ${LDFLAGS} --entry=${ENTRY_NAME} -T${LINKER_SCRIPT} <OBJECTS> ${STARTUP_OBJS} -o <TARGET> <LINK_LIBRARIES> -L${NUTTX_PATH}/libs --start-group ${LDLIBS} ${EXTRA_LIBS} --end-group"
+    "<CMAKE_LINKER> ${LDFLAGS} --entry=${ENTRY_NAME} -T${LINKER_SCRIPT} <OBJECTS> -o <TARGET> <LINK_LIBRARIES> -L${NUTTX_PATH}/libs --start-group ${LDLIBS} ${EXTRA_LIBS} --end-group"
 )
 
 set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)

